{"cells":[{"cell_type":"markdown","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"f1b528de2aec1ac63ed6a12f47afae79","grade":false,"grade_id":"cell-84f9f7a5056b3805","locked":true,"schema_version":3,"solution":false,"task":false},"tags":[],"id":"463UJsebD9CK"},"source":[" # COMP90014 Assignment 3\n","### Semester 2, 2024\n","Version 1.0 Last edited 25/09/2024."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"3M_z1UdRD9CM"},"outputs":[],"source":["### Fill in your student details here\n","NAME = \"Ching-Yi Peng\"\n","ID = \"1545824\""]},{"cell_type":"markdown","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"d061a98501a48489971937a235656b33","grade":false,"grade_id":"cell-745a48f0f8f3fafd","locked":true,"schema_version":3,"solution":false,"task":false},"tags":[],"id":"vV2UhkGlD9CN"},"source":["## Completing the assignment\n","\n","***Academic integrity***\n","\n","This assignment should be completed by each student individually. <br>\n","Make sure you read this entire document, and ask for help if anything is not clear. <br>\n","Any changes or clarifications to this document will be announced via the LMS. <br>\n","\n","Please make sure you review the University's rules on academic honesty and plagiarism: https://academichonesty.unimelb.edu.au/\n","\n","Do not copy any code from other students or from the internet. This is considered plagiarism.\n","\n","***Completing the assignment & submission***\n","\n","To complete the assignment, finish the coding and short answer tasks in this notebook.<br>\n","You are free to use any IDE to work on the assignment, but only use jupyter to prepare your final submission (ensures compatability with grading software). <br>\n","Please do not copy or delete cells in this notebook, as this may interrupt the autograding of hidden tests. <br>\n","Your completed notebook file containing all your answers will be turned in via LMS. Please also submit an HTML file.\n","\n","***Visible & hidden tests***\n","\n","In some cases, we have provided test input and test output that you can use to try out your solutions. <br>\n","These are visible tests and serve to warn you if you've made a mistake but are **not** exhaustive.\n","\n","During assessment, there are also hidden tests we run to validate your code. <br>\n","As you won't see the hidden tests, it's up to you to decide whether your code is correct.\n","\n","**Remember to save your work early and often.**"]},{"cell_type":"markdown","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"5b0ae1b4c5b2501411a6284d2690e370","grade":false,"grade_id":"cell-cf0f1f5a83489bff","locked":true,"schema_version":3,"solution":false,"task":false},"tags":[],"id":"TzuSYYUOD9CO"},"source":["## Marking\n","\n","***Graded cells***\n","\n","Cells that must be completed to receive marks are clearly labelled with the following text:\n","\n","`# -- GRADED CELL (N marks) - complete this cell --`\n","\n","Only add answers to these cells.\n","\n","Some cells are code cells, in which you must complete the code to solve a problem. <br>\n","Others are markdown cells, in which you must write your answers to short-answer questions.\n","\n","***Completing code cells***\n","\n","You will see the following text in graded code cells.\n","\n","``` python\n","raise NotImplementedError()\n","```\n","\n","You must remove the `raise NotImplementedError()` line from the cell, and replace it with your solution. <br>\n","If you want to import a library or use a helper function, the import statements must be inside the function for that question. <br>\n","For example:\n","    \n","``` python\n","# -- GRADED CELL (1 mark) - complete this cell --\n","\n","def question1(param1: str, param2: int) -> float:\n","    from collections import defaultdict\n","```\n","\n","Include code comments in your solutions! <br>\n","Well commented code can help you to receive partial marks even if the final solution is incorrect.\n","\n","***Editing the notebook***\n","\n","Only graded cells will be marked.\n","- Do **NOT** enter solutions outside of graded cells\n","- Do **NOT** duplicate or remove cells from the notebook\n","- You may add new cells to test code, but new cells will not be graded.\n","- Word limits, where stated, will be strictly enforced. Answers exceeding the limit **will not be marked**.\n","\n","***Marks***\n","\n","The total marks for the assignment add up to 15. <br>\n","This assignment will be worth 15% of your overall subject grade.\n","\n"]},{"cell_type":"markdown","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"74ec1f370172303aa1ab8b787f67a1a2","grade":false,"grade_id":"cell-1fa2695595062ed1","locked":true,"schema_version":3,"solution":false,"task":false},"tags":[],"id":"yVZDGWJ-D9CO"},"source":["## Submission\n","\n","Only use jupyter to prepare your final submission. This ensures compatability with our grading software. <br>\n","\n","Make sure you have filled in any place that says `# -- GRADED CELL (N marks) - complete this cell --`.\n","\n","Before you turn this assignment in, make sure everything runs as expected, and the output is cleared.\n","\n","First, **restart the kernel**:\n","\n","- In the menubar, select Kernel -> Restart.\n","\n","Next, **run all cells**:\n","\n","- In the menubar, select Cell -> Run All.\n","\n","Finally, **clear all output**:\n","\n","- In the menubar, select Options -> Clear All Outputs\n","\n","Your completed notebook file containing all your answers must be turned in via LMS in `.ipynb` format. <br>\n","You must also submit a copy of this notebook in `html` format with the output cleared.\n","\n","Your submission should include **only two** files with names formatted as: **Assignment3.ipynb** and **Assignment3.html**\n","\n"]},{"cell_type":"markdown","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"70e7cec4657075e1a167b942db7f4062","grade":false,"grade_id":"cell-529f2989ec65a759","locked":true,"schema_version":3,"solution":false,"task":false},"tags":[],"id":"7Fg5Ox_GD9CO"},"source":["<div style=\"background: rgb(255,165,0); border: solid 1px rgb(129,199,132); padding: 10px;\">    \n","\n","<h1>Assignment 3</h1>\n","\n","<h2>De Bruijn Graph Simplification</h2>\n","\n","</div>"]},{"cell_type":"markdown","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"4cc8c70fac77e8e94abdc84305b79b2d","grade":false,"grade_id":"cell-9c57d397c05bc46c","locked":true,"schema_version":3,"solution":false,"task":false},"tags":[],"id":"rbXKvheeD9CP"},"source":["**Structure**\n","\n","This assignment has two sections.\n","\n","- In section 1 we will implement a *modified version* of the tour bus algorithm, inspired from the [Velvet](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2336801/) paper.<br>\n","- In section 2 you need to answer two short-answer questions related to dimensionality reduction and clustering.\n","\n","**Section 1: The Tour Bus [10 marks]**\n","- Q1: explore [2 marks]\n","- Q2: backtrack [2 marks]\n","- Q3: compare [2 marks]\n","- Q4: select_consensus [2 marks]\n","- Q5: collapse [2 marks]\n","- <span style=\"color: blue\">Q6: collapse_hard [+3 bonus marks] (optional)</span>\n","\n","**Section 2: Short-answer questions [5 marks]**\n","- Q7: Dimensionality Reduction [2 marks]\n","- Q8: Clustering [3 marks]\n"]},{"cell_type":"markdown","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"0a17ccbaa3378ec71af2e205fe484782","grade":false,"grade_id":"cell-4c3fb5ef1811ec08","locked":true,"schema_version":3,"solution":false,"task":false},"tags":[],"id":"ETlDi-cGD9CP"},"source":["**Packages**\n","\n","This assignment uses the <small>`networkx`</small> and <small>`python-levenshtein`</small> packages. <br>\n","You may use functionality within <small>`networkx`</small> to answer questions in this assignment (if helpful). <br>\n","These are both installable via pip and conda.\n"]},{"cell_type":"code","source":["!pip install python-Levenshtein"],"metadata":{"id":"eo-HWp-obIV7"},"execution_count":null,"outputs":[]},{"cell_type":"code","execution_count":null,"metadata":{"tags":[],"id":"xHKWkZLlD9CQ"},"outputs":[],"source":["# imports\n","import networkx as nx\n","import Levenshtein\n","from typing import Tuple, List, Optional"]},{"cell_type":"markdown","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"254c9ad8ccff70edd8cc514d7d063343","grade":false,"grade_id":"cell-0bab95cb051cd51a","locked":true,"schema_version":3,"solution":false,"task":false},"tags":[],"id":"DM8mtgpZD9CQ"},"source":["**Data**\n","\n","The following files are required for this assignment:\n","- https://github.com/melbournebioinformatics/COMP90014_2024/raw/master/assignments/data/nodelist.tsv\n","- https://github.com/melbournebioinformatics/COMP90014_2024/raw/master/assignments/data/edgelist.tsv\n","- https://github.com/melbournebioinformatics/COMP90014_2024/raw/master/assignments/data/kmer_counts.tsv\n","\n","Download the files above and place them in a folder named 'data' relative to your current working directory. <br>\n","Run the cell below to confirm the paths are correct."]},{"cell_type":"code","source":["!mkdir -p data\n","\n","!wget -P data https://github.com/melbournebioinformatics/COMP90014_2024/raw/master/assignments/data/nodelist.tsv\n","!wget -P data https://github.com/melbournebioinformatics/COMP90014_2024/raw/master/assignments/data/edgelist.tsv\n","!wget -P data https://github.com/melbournebioinformatics/COMP90014_2024/raw/master/assignments/data/kmer_counts.tsv"],"metadata":{"id":"Jqe6rBfBbdM5"},"execution_count":null,"outputs":[]},{"cell_type":"code","execution_count":null,"metadata":{"tags":[],"id":"L6-QzbUjD9CQ"},"outputs":[],"source":["# NOTE: DO NOT CHANGE THESE PATHS.\n","# Changing these variables will result in a 1 mark deduction to your grade.\n","NODELIST = 'data/nodelist.tsv'\n","EDGELIST = 'data/edgelist.tsv'\n","KMERCOUNTS = 'data/kmer_counts.tsv'\n","\n","import os\n","for filepath in [NODELIST, EDGELIST, KMERCOUNTS]:\n","    if os.path.exists(filepath):\n","        print(f'found {filepath}')\n","    else:\n","        print(f'error {filepath}')\n"]},{"cell_type":"markdown","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"c6b96327512393fab54cc07376dc282d","grade":false,"grade_id":"cell-023a328cb031702b","locked":true,"schema_version":3,"solution":false,"task":false},"tags":[],"id":"-dndRMRhD9CQ"},"source":["The graphs used in visible test cases has the same structure as appears in figures.  <br>\n","A different graph is used for hidden tests. You are encouraged to write new tests so that your code covers edge-cases not present in the visible tests.\n","\n","Run the cells below to load the nx.DiGraph we will be working with in this assignment.\n"]},{"cell_type":"code","execution_count":null,"metadata":{"tags":[],"id":"7GhxkCiVD9CR"},"outputs":[],"source":["# helper function to load graph from file.\n","def load_graph(nodelist_path: str, edgelist_path: str) -> nx.DiGraph:\n","    G = nx.DiGraph()\n","\n","    # read in node data\n","    with open('data/nodelist.tsv', 'r') as fp:\n","        lines = fp.readlines()\n","        lines = [ln.strip().split('\\t') for ln in lines]\n","        for node, seq in lines[1:]:\n","            G.add_node(node, sequence=seq)\n","\n","    # read in edge data\n","    with open('data/edgelist.tsv', 'r') as fp:\n","        lines = fp.readlines()\n","        lines = [ln.strip().split('\\t') for ln in lines]\n","        for src, dest, kmer in lines[1:]:\n","            G.add_edge(src, dest, kmer=kmer)\n","\n","    # return graph\n","    return G\n"]},{"cell_type":"code","execution_count":null,"metadata":{"tags":[],"id":"oP34eXfSD9CR"},"outputs":[],"source":["# viewing the graph data\n","G = load_graph(NODELIST, EDGELIST)\n","\n","print('\\nNODES ---')\n","print('node\\tsequence')\n","for node, data in G.nodes(data=True):\n","    print(node + '\\t' + data['sequence'])\n","\n","print('\\nEDGES ---')\n","print('node1\\tnode2\\tkmer')\n","for n1, n2, data in G.edges(data=True):\n","    print(n1 + '\\t' + n2 + '\\t' + data['kmer'])"]},{"cell_type":"markdown","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"ce0be6f48eb338cd3ca58d691d67121e","grade":false,"grade_id":"cell-498be4ab3159b8dd","locked":true,"schema_version":3,"solution":false,"task":false},"tags":[],"id":"J8z0h0Y-D9CR"},"source":["<br>\n","\n","<div style=\"color: rgb(0,0,0); background: #0096FF; border: solid 1px rgb(0,0,0); padding: 10px;\">\n","\n","<h1>Section 1: The Tour Bus</h1>\n","\n","</div>\n"]},{"cell_type":"markdown","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"2defaf9a94627db0bb603b1eb6686375","grade":false,"grade_id":"cell-93aa737b463692da","locked":true,"schema_version":3,"solution":false,"task":false},"tags":[],"id":"4WwqcZxyD9CR"},"source":["### Introduction\n","\n","Complex regions in De Bruijn Graphs can be caused by many factors, including read errors, heterozygosity, and repetitive sequence.\n","To handle such situations, the genome assembler 'Velvet' uses an approach called the 'Tour Bus'.\n","\n","The main ideas of the Tour Bus are as follows:\n","- Two branching paths, which start and end at the same node, are termed 'bubbles' due to their topological appearence.\n","- These paths may encode similar sequence along their length, and if so, may have formed due to sequencing errors or biological variability (eg heterozygosity).  \n","- In such a situation, these paths can be collapsed into a single 'consensus' path, simplifying the graph (and improving the final assembly).\n","\n","<img src=\"https://raw.githubusercontent.com/melbournebioinformatics/COMP90014_2024/master/assignments/media/tourbus.png\" width=\"600\">\n","\n","In the figure above we see an example of this situation. <br>\n","The 'consensus' path is the series of nodes [A, B, C, D, E], with nodes B', C', C'', and D' being caused by the types of errors mentioned above.\n","\n","You will notice that there are two bubbles in this graph. <br>\n","As such, bubbles need to be handled iteratively, where we detect and resolve a single bubble at each iteration.\n","\n"]},{"cell_type":"markdown","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"945ac2cfb2ab06767bedf675092cfaea","grade":false,"grade_id":"cell-5ac2b19f155c1d93","locked":true,"schema_version":3,"solution":false,"task":false},"tags":[],"id":"h1kRz3FoD9CS"},"source":["<br>\n","\n","-----------------\n","\n","### Identifying Bubbles\n","\n","The first step of each Tour Bus iteration is to identify whether a bubble exists. <br>\n","The graph is explored, starting at an arbitrary node, and stopping when a previously visited node is a neighbour of the current node. <br>\n","In this situation, both the current node and the visited node are returned.\n","\n","<img src=\"https://raw.githubusercontent.com/melbournebioinformatics/COMP90014_2024/master/assignments/media/explore.png\" width=\"600\">\n","\n"]},{"cell_type":"markdown","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"d1847457ab553b6a45b1396fdadda104","grade":false,"grade_id":"cell-46098e7111c9793e","locked":true,"schema_version":3,"solution":false,"task":false},"tags":[],"id":"4WWx91O6D9CS"},"source":["<div style=\"color: rgb(27,94,32); background: rgb(200,230,201); border: solid 1px rgb(129,199,132); padding: 10px;\">\n","<h3>Question 1</h3>\n","\n","<b>Challenge:</b> Write a function which explores a graph to identify bubbles. <br>\n","\n","- [ ] Input: G (networkx.DiGraph); node (str).\n","- [ ] Ouput: A tuple with the format (current node, visited node).\n","- [ ] If there is no bubble to resolve, return <small>`None`</small>.\n","\n","When exploring node neighbours, break ties alphabetically. <br>\n","For example, if the neighbours of node <small>`A`</small> are <small>`B`</small> and <small>`B'`</small>, you should visit <small>`B`</small> first.\n","    \n","</div>\n"]},{"cell_type":"code","execution_count":null,"metadata":{"deletable":false,"nbgrader":{"cell_type":"code","checksum":"cb260db6b47dbecf385b4c27281e99df","grade":false,"grade_id":"cell-a9737404fb0ce832","locked":false,"schema_version":3,"solution":true,"task":false},"tags":[],"id":"SCna_3Q6D9CS"},"outputs":[],"source":["# -- GRADED CELL (2 marks) - complete this cell --\n","from collections import deque\n","\n","def explore(G: nx.DiGraph, node: str) -> Optional[Tuple]:\n","    \"\"\"\n","    Explores the graph from a provided starting node to identify a bubble.\n","    Returns the current node and previously visited node as a tuple if present.\n","    If no bubble is identified, returns None.\n","    \"\"\"\n","    visited = set()\n","    queue = deque(node)\n","\n","    while queue:\n","        current_node = queue.popleft()\n","        visited.add(current_node)\n","\n","        # Get neighbors of the current node sorted alphabetically\n","        neighbors = sorted(G.neighbors(current_node))\n","\n","        for neighbor in neighbors:\n","            if not neighbor in visited:\n","                queue.append(neighbor)\n","                visited.add(neighbor)\n","            else:\n","              return (current_node, neighbor)\n","\n","    # If no bubble is found\n","    return None\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"Er-eTWKiD9CT"},"outputs":[],"source":["# extra code cell for development if needed\n"]},{"cell_type":"code","execution_count":null,"metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"code","checksum":"918f93e661897aadaddcc421a635ded2","grade":true,"grade_id":"cell-4cfada36fef96a3b","locked":true,"points":2,"schema_version":3,"solution":false,"task":false},"tags":[],"id":"7MBl8OttD9CT"},"outputs":[],"source":["# Testing cell - DO NOT ALTER.\n","G = load_graph(NODELIST, EDGELIST)\n","\n","# test1: from node A\n","res = explore(G, 'A')\n","print(f\"test1: expected=('C`', 'D'), actual={res}\")\n","\n","# test2: from node B\n","res = explore(G, 'B')\n","print(f\"\\ntest2: expected=('D', 'E'), actual={res}\")\n","\n","# test3: from node C\n","res = explore(G, 'C')\n","print(f\"\\ntest3: expected=None, actual={res}\")\n"]},{"cell_type":"markdown","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"934e008f92b0b4371704499994655ecb","grade":false,"grade_id":"cell-5b2dcefefffa2ade","locked":true,"schema_version":3,"solution":false,"task":false},"tags":[],"id":"Pet_bIA0D9CT"},"source":["<br>\n","\n","--------------------\n","\n","### Backtracking\n","\n","After a bubble has been identified, the second step is to extract the two paths involved in the bubble.\n","\n","<img src=\"https://raw.githubusercontent.com/melbournebioinformatics/COMP90014_2024/master/assignments/media/backtrack.png\" width=\"600\">"]},{"cell_type":"markdown","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"730fef14d132ad3dee951253d3d2d548","grade":false,"grade_id":"cell-c327c40f0ce12fb7","locked":true,"schema_version":3,"solution":false,"task":false},"tags":[],"id":"q5QGTq8nD9CU"},"source":["<div style=\"color: rgb(27,94,32); background: rgb(200,230,201); border: solid 1px rgb(129,199,132); padding: 10px;\">\n","<h3>Question 2</h3>\n","\n","<b>Challenge:</b> Write a function which uses backtracking to return two paths involved in a bubble.<br>\n","\n","- [ ] Input: G (networkx.DiGraph); current_node (str); visited_node (str).\n","- [ ] Ouput: A tuple with the format (path1, path2), where each path is a list of nodes.\n","\n","Details:\n","- Backtrack from <small>`current_node`</small> and <small>`visited_node`</small>.\n","- <small>`visited_node`</small> should appear in both output paths, so that both paths start and end at the same node.\n","- Ensure output paths read left to right in the figure above. ie node <small>`A`</small> should be the first item, and node <small>`D`</small> should be the last item.\n","\n","Consider situations where a node has multiple in-edges, for example node <small>`B'`</small> in the figure above.\n","    \n","</div>\n"]},{"cell_type":"code","execution_count":null,"metadata":{"deletable":false,"nbgrader":{"cell_type":"code","checksum":"8739d63ca57c34d09d55690f6bc96d4e","grade":false,"grade_id":"cell-3783c07ccafbcd68","locked":false,"schema_version":3,"solution":true,"task":false},"tags":[],"id":"qiX_OK8MD9CU"},"outputs":[],"source":["# -- GRADED CELL (2 marks) - complete this cell --\n","\n","def backtrack(G: nx.DiGraph, current_node: str, visited_node: str) -> Tuple[List, List]:\n","    \"\"\"\n","    Performs backtracking to extract two paths (of nodes) involved in a bubble.\n","    \"\"\"\n","    path1 = [current_node, visited_node]\n","    path2 = [visited_node]\n","    # Track visited nodes to avoid cycles\n","    visited = set([current_node])\n","\n","    # Backtrack from current_node to visited_node to find path1\n","    while current_node != visited_node:\n","        predecessors = sorted(G.predecessors(current_node))\n","        found_predecessor = False\n","        for pred in predecessors:\n","            if pred not in visited:\n","                path1.append(pred)\n","                current_node = pred\n","                visited.add(pred)\n","                found_predecessor = True\n","                break\n","        # If no predecessor is found, break the loop to avoid infinite loops\n","        if not found_predecessor:\n","            break\n","\n","    # Reset current_node to visited_node for the second path\n","    current_node = visited_node\n","    visited = set([current_node])\n","\n","    # Backtrack to find an alternate path (path2)\n","    while current_node != path1[-1]:\n","        predecessors = sorted(G.predecessors(current_node))\n","        found_predecessor = False\n","        for pred in predecessors:\n","            # Avoid using predecessors already used in path1\n","            if pred not in visited:\n","                path2.append(pred)\n","                current_node = pred\n","                visited.add(pred)\n","                found_predecessor = True\n","                break\n","        if current_node in path1:\n","            # current_index = path1.index(current_node)\n","            # if current_index > 0:\n","            #     previous_node = path1[current_index]\n","            #     path1.remove(previous_node)\n","            break\n","        # If no predecessor is found, break the loop to avoid infinite loops\n","        if not found_predecessor:\n","            break\n","\n","    # Reverse paths to maintain left-to-right order\n","    return (sorted(path1), sorted(path2))\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"GJU5CF2mD9CU"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"code","checksum":"ff99ff65ea90bb648aa5e38eecc33e8c","grade":true,"grade_id":"cell-fb437821dcaa2960","locked":true,"points":2,"schema_version":3,"solution":false,"task":false},"tags":[],"id":"B3X8_ULqD9CU"},"outputs":[],"source":["# Testing cell - DO NOT ALTER.\n","G = load_graph(NODELIST, EDGELIST)\n","\n","# test1: as in figure\n","path1, path2 = backtrack(G, 'C`', 'D')\n","print('test 1')\n","print(f\"expected: path1 = ['A', 'B`', 'C`', 'D'], path2 = ['A', 'B', 'C', 'D']\")\n","print(f\"actual:   path1 = {path1}, path2 = {path2}\")\n","\n","# another bubble\n","path1, path2 = backtrack(G, 'D', 'E')\n","print('\\ntest 2')\n","print(f\"expected: path1 = ['B', 'C', 'D', 'E'], path2 = ['B', 'C``D`', 'E']\")\n","print(f\"actual:   path = {path1}, path2 = {path2}\")\n"]},{"cell_type":"markdown","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"d820c10df4d4e185b42ff3cf940de682","grade":false,"grade_id":"cell-776a1c162b760f2d","locked":true,"schema_version":3,"solution":false,"task":false},"tags":[],"id":"T14LqoIED9CU"},"source":["<br>\n","\n","----------------------------\n","\n","### Comparing sequences\n","\n","Now that a bubble has been identified and two node paths have been fetched, we need to check how similar the sequences are along these paths.<br>\n","If highly similar (determined using levenshtein distance), we can confidently assume the bubble was due to biological variation or sequencing error.<br>\n","If too dissimilar, we should ignore the bubble.\n","\n","The figure below demonstrates two situations you will need to handle, as the two paths may not have the same number of nodes.\n","\n","<img src=\"https://raw.githubusercontent.com/melbournebioinformatics/COMP90014_2024/master/assignments/media/compare.png\" width=\"500\">"]},{"cell_type":"markdown","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"16a0de723114545ccf129d13522ff19b","grade":false,"grade_id":"cell-bd35e5a7423ad46f","locked":true,"schema_version":3,"solution":false,"task":false},"tags":[],"id":"jlEWA5l2D9CV"},"source":["<div style=\"color: rgb(27,94,32); background: rgb(200,230,201); border: solid 1px rgb(129,199,132); padding: 10px;\">\n","<h3>Question 3</h3>\n","\n","<b>Challenge:</b> Write a function <small>`compare()`</small> which compares the sequence along two paths to decide whether the paths can be merged. <br>\n","\n","- [ ] Input: G (networkx.DiGraph); path1 (list of str); path2 (list of str); a kmer-size k (int) (which was originally used to generate the De Bruijn Graph).\n","- [ ] Ouput: True or False (bool).\n","\n","Details:\n","- Think about whether the start and end node of each path (shared) should be considered.\n","- Consider that nodes in De Buijn Graphs are originally <small>`k-1`</small> prefix/suffix pairs, and how this will affect the sequence along each path.\n","- The sequence for a given node is available via a data attribute named <small>`'sequence'`</small> (eg <small>`G.nodes['A']['sequence']`</small>)\n","- A function named <small>`get_edits()`</small> has been provided below to return the number of edits between two sequences.\n","- Decision condition - the number of edits must be less-than or equal-to 20% of the longer sequence for the paths to be considered 'similar'.\n","    \n","</div>\n"]},{"cell_type":"code","execution_count":null,"metadata":{"deletable":false,"nbgrader":{"cell_type":"code","checksum":"3660fba803849e7ddec907a90e83818a","grade":false,"grade_id":"cell-f238a10ac0b8b9d6","locked":false,"schema_version":3,"solution":true,"task":false},"tags":[],"id":"rNjIIbXTD9CV"},"outputs":[],"source":["# -- GRADED CELL (2 marks) - complete this cell --\n","\n","# helper function\n","def do_align(seq1: str, seq2: str) -> float:\n","    return Levenshtein.distance(seq1, seq2)\n","\n","def compare(G: nx.DiGraph, path1: List[str], path2: List[str], k: int=5) -> bool:\n","    \"\"\"\n","    Compares the sequence along two paths to decide whether they are similar enough to merge.\n","    The number of edits must be less than 20% of the length of the longer sequence to be considered as 'similar'.\n","    \"\"\"\n","    # Concatenate the sequences of all nodes in path1\n","    full_seq1 = \"\".join([G.nodes[node]['sequence'] for node in path1])\n","\n","    # Concatenate the sequences of all nodes in path2\n","    full_seq2 = \"\".join([G.nodes[node]['sequence'] for node in path2])\n","\n","    # Compute the Levenshtein distance between the full sequences\n","    edits = do_align(full_seq1, full_seq2)\n","\n","    # Get the length of the longer sequence\n","    max_length = max(len(full_seq1), len(full_seq2))\n","\n","    return edits <= 0.2 * max_length\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"_utB_DgAD9CV"},"outputs":[],"source":["# extra code cell for development if needed\n"]},{"cell_type":"code","execution_count":null,"metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"code","checksum":"2fef07c5f4e4433703b0196fa16ac379","grade":true,"grade_id":"cell-d9b2ea12560807dd","locked":true,"points":2,"schema_version":3,"solution":false,"task":false},"tags":[],"id":"EcdGFos6D9CV"},"outputs":[],"source":["# Testing cell - DO NOT ALTER.\n","G = load_graph(NODELIST, EDGELIST)\n","\n","# test1: a valid bubble\n","res = compare(G, ['A', 'B`', 'C`', 'D'], ['A', 'B', 'C', 'D'], k=5)\n","print(f'test1: expected=True, actual={res}')\n","\n","# test2: another valid bubble\n","print()\n","res = compare(G, ['B', 'C', 'D', 'E'], ['B', 'C``D`', 'E'], k=5)\n","print(f'test2: expected=True, actual={res}')\n","\n","# test3: not a bubble (or valid paths for that matter), but valid as a test case.\n","print()\n","res = compare(G, ['B`', 'C`', 'D'], ['B`', 'X', 'D'], k=5)\n","print(f'test3: expected=False, actual={res}')\n"]},{"cell_type":"markdown","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"b1a90f02931a60371abd27eae862917e","grade":false,"grade_id":"cell-b8c11aba7bb566b7","locked":true,"schema_version":3,"solution":false,"task":false},"tags":[],"id":"uz8QSMBRD9CV"},"source":["<br>\n","\n","---------------------------\n","\n","### Selecting the consensus (correct) path\n","\n","Once the two paths have been deemed similar enough to merge, we can collapse the bubble.\n","\n","To begin, the 'consensus' path must be identified. <br>\n","This indicates which path is likely correct, and is calculated using kmer counts from the original readset used to construct our De Bruijn Graph. <br>\n","\n","The De Bruijn Graph we are working with has already been simplified by coalescing linear chains of nodes. <br>\n","This transformation can be seen in the figure below.\n","\n","<img src=\"https://raw.githubusercontent.com/melbournebioinformatics/COMP90014_2024/master/assignments/media/consensus.png\" width=\"600\">\n"]},{"cell_type":"markdown","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"0516eccfbc6d7bcc2cd2544f98a73cbb","grade":false,"grade_id":"cell-0b420856114cd573","locked":true,"schema_version":3,"solution":false,"task":false},"tags":[],"id":"380rvCxxD9CV"},"source":["To determine the consensus path, we will calculate a 'path length' for each path in the bubble. <br>\n","The formula we will use is <small>`len(p) / multiplicity(p)`</small>, where:\n","- <small>`len(p)`</small> refers to the number of kmers along the path\n","- <small>`multiplicity(p)`</small> refers to the cumulative counts (multiplicity) of kmers along the path.\n","\n","These kmer counts are available via the provided <small>`kmer_counts`</small> dictionary which maps kmers to their associated multiplicity (represented as a table in the figure above).<br>\n","As displayed in the figure above, the path length remains unchanged after linear chains have been coalesced, as kmers can be extracted out of node sequences. <br>\n","\n","For example:\n","- The path ['GGAT', 'GATTACA', 'ACAT'] will have a path length of <small>`5 / (10+12+10+11+9) = 0.096`</small>.\n","- The path ['CGAT', 'GATTACA', 'ACAA'] will have a path length of <small>`5 / (1+12+10+11+2) = 0.139`</small>.\n","\n","The path with the smaller 'path length' will be selected as the consensus, as prioritises the path with stronger read evidence.\n"]},{"cell_type":"markdown","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"80ad5f54dafd3d7c643f949712079beb","grade":false,"grade_id":"cell-2b4100ea82b2480d","locked":true,"schema_version":3,"solution":false,"task":false},"tags":[],"id":"fCNeEZO4D9CV"},"source":["<div style=\"color: rgb(27,94,32); background: rgb(200,230,201); border: solid 1px rgb(129,199,132); padding: 10px;\">\n","<h3>Question 4</h3>\n","\n","<b>Challenge:</b> Write a function which identifies the consensus path from two candidate paths. <br>\n","\n","- [ ] Input: G (networkx.DiGraph); path1 (list of str); path2 (list of str); kmer_counts (dict); a kmer size k (int).\n","- [ ] Ouput: One of the input paths (the consensus path).\n","    \n","Edge kmers are available via a data attribute named <small>`'kmer'`</small> for a given edge (eg <small>`G.edges['A', 'B']['kmer']`</small>)\n","\n","</div>\n"]},{"cell_type":"code","execution_count":null,"metadata":{"deletable":false,"nbgrader":{"cell_type":"code","checksum":"33502686b6979b328e8b53f4da93dec7","grade":false,"grade_id":"cell-549cee9515f400b9","locked":false,"schema_version":3,"solution":true,"task":false},"tags":[],"id":"2x4Y-pzdD9CV"},"outputs":[],"source":["# -- GRADED CELL (2 marks) - complete this cell --\n","\n","def select_consensus(G: nx.DiGraph, path1: List[str], path2: List[str], kmer_counts: dict, k: int=5) -> List:\n","    \"\"\"\n","    Selects and returns the consensus path by calculating the path with lower 'path length'.\n","    \"\"\"\n","    def calculate_path_length(path):\n","        total_kmer_count = 0\n","        for i in range(len(path) - 1):\n","            edge_kmer = G.edges[path[i], path[i + 1]].get('kmer')\n","            if edge_kmer is None:\n","                continue  # Skip this edge if no k-mer is found\n","\n","            kmer_count = kmer_counts.get(edge_kmer, 0)\n","            total_kmer_count += kmer_count\n","\n","        # Avoid division by zero in case of missing k-mer counts\n","        if total_kmer_count == 0:\n","            return float('inf')\n","\n","        # Calculate path length using the formula\n","        path_length = k / total_kmer_count\n","        return path_length\n","\n","    # Calculate the path lengths for both paths\n","    path1_length = calculate_path_length(path1)\n","    path2_length = calculate_path_length(path2)\n","\n","    # Select the path with the smaller path length\n","    if path1_length < path2_length:\n","        return path1\n","    elif path2_length < path1_length:\n","        return path2\n","    else:\n","        return None\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"c97gKjleD9CW"},"outputs":[],"source":["# extra code cell for development if needed\n"]},{"cell_type":"code","execution_count":null,"metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"code","checksum":"cfb446df211f8c74a0e7209af9a14839","grade":true,"grade_id":"cell-cf9103140f93a4b9","locked":true,"points":2,"schema_version":3,"solution":false,"task":false},"tags":[],"id":"uMq8FDTpD9CW"},"outputs":[],"source":["# Testing cell - DO NOT ALTER.\n","G = load_graph(NODELIST, EDGELIST)\n","with open(KMERCOUNTS, 'r') as fp:\n","    lines = fp.readlines()\n","    lines = [ln.strip().split('\\t') for ln in lines[1:]]\n","    kmer_counts = {k: int(c) for k, c in lines}\n","\n","# single test: a valid bubble\n","print()\n","path1 = ['A', 'B`', 'C`', 'D']\n","path2 = ['A', 'B', 'C', 'D']\n","res = select_consensus(G, path1, path2, kmer_counts, k=5)\n","print('note: the path length of path1 should be approximately 0.2, and path length of path2 should be approximately 0.1')\n","print(f\"test1: expected=['A', 'B', 'C', 'D'], actual={res}\")\n","\n"]},{"cell_type":"markdown","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"5ddb39cc8e95380c2737d725063204bf","grade":false,"grade_id":"cell-cdede2b9a1bb1711","locked":true,"schema_version":3,"solution":false,"task":false},"tags":[],"id":"k0ostwFrD9CW"},"source":["<br>\n","\n","----------------\n","\n","### Collapsing the Bubble\n","\n","Finally, we can collapse the bubble. <br>\n","We will take an iterative approach as displayed in the figure below.\n","\n","Note that the edge <small>`B' -> X`</small> is regrafted to <small>`B -> X`</small>.<br>\n","This somewhat violates the concept of edges in a De Bruijn Graph being kmers which connect prefix/suffix pairs, but is required nonetheless.\n","\n","<img src=\"https://raw.githubusercontent.com/melbournebioinformatics/COMP90014_2024/master/assignments/media/collapse.png\" width=\"500\">\n","\n"]},{"cell_type":"markdown","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"f76ba7b096b0898c958c40275d9ffe12","grade":false,"grade_id":"cell-2475487398fec1f9","locked":true,"schema_version":3,"solution":false,"task":false},"tags":[],"id":"T3-CeB1mD9CW"},"source":["<div style=\"color: rgb(27,94,32); background: rgb(200,230,201); border: solid 1px rgb(129,199,132); padding: 10px;\">\n","<h3>Question 5</h3>\n","\n","<b>Challenge:</b> Write a function which collapses the graph bubble. <br>\n","\n","- [ ] Input: G (networkx.DiGraph); consensus_path (list of str); minority_path (list of str).\n","- [ ] Ouput: Updated graph with the bubble removed (networkx.DiGraph)\n","\n","</div>\n"]},{"cell_type":"code","execution_count":null,"metadata":{"deletable":false,"nbgrader":{"cell_type":"code","checksum":"cd2972532dbbcd2425603cdb4ca194af","grade":false,"grade_id":"cell-c9785e6c5122d3be","locked":false,"schema_version":3,"solution":true,"task":false},"tags":[],"id":"-PH_PP1HD9CW"},"outputs":[],"source":["# -- GRADED CELL (2 marks) - complete this cell --\n","\n","def collapse(G: nx.DiGraph, consensus_path: List[str], minority_path: List[str]) -> nx.DiGraph:\n","    \"\"\"\n","    Collapses a bubble by removing the minority_path from the graph.\n","    Returns the simplified graph G with bubble removed.\n","    \"\"\"\n","    for node in minority_path:\n","        # Only remove the node if it's not part of the consensus path\n","        if node not in consensus_path:\n","            if G.has_node(node):\n","                G.remove_node(node)\n","\n","    return G"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"PAnuZbK1D9CW"},"outputs":[],"source":["# extra code cell for development if needed\n","\n","def collapse_2(G: nx.DiGraph, consensus_path: List[str], minority_path: List[str]) -> nx.DiGraph:\n","    \"\"\"\n","    Collapses a bubble by removing the minority_path from the graph.\n","    Returns the simplified graph G with bubble removed.\n","    \"\"\"\n","    for node_co in consensus_path:\n","        for node_min in minority_path:\n","            if node_co != node_min and G.has_node(node_min):\n","                G.remove_node(node_min)  # Remove nodes in the minority path\n","\n","        # Check if the current consensus node exists before accessing its predecessors\n","        if G.has_node(node_co):\n","            pre_nodes = list(G.predecessors(node_co))\n","\n","            # For each predecessor, contract it with the current node in the consensus path\n","            for pre in pre_nodes:\n","                if G.has_node(pre):\n","                    # Contract predecessor into the current consensus node\n","                    G = nx.contracted_nodes(G, node_co, pre, self_loops=False)\n","\n","    return G"]},{"cell_type":"code","execution_count":null,"metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"code","checksum":"cb705f769df61cab94064a8da104e752","grade":true,"grade_id":"cell-ef77d1b1626875e2","locked":true,"points":2,"schema_version":3,"solution":false,"task":false},"tags":[],"id":"HIgxGpSRD9CX"},"outputs":[],"source":["# Testing cell - DO NOT ALTER.\n","G = load_graph(NODELIST, EDGELIST)\n","\n","# test1: the first bubble\n","consensus_path = ['A', 'B', 'C', 'D']\n","minority_path = ['A', 'B`', 'C`', 'D']\n","G = collapse(G, consensus_path, minority_path)\n","print(\"\\ntest1 (remaining nodes)\")\n","print(\"expected: [['AB', 'C', 'C``D`', 'D', 'E', 'X']\")\n","print(f\"actual:   {sorted(G.nodes())}\")\n","\n","# test2: the second bubble\n","consensus_path = ['AB', 'C', 'D', 'E']\n","minority_path = ['AB', 'C``D`', 'E']\n","G = collapse(G, consensus_path, minority_path)\n","print(\"\\ntest2 (remaining nodes)\")\n","print(\"expected: ['AB', 'CDE', 'X']\")\n","print(f\"actual:   {sorted(G.nodes())}\")\n"]},{"cell_type":"markdown","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"1749e95f827a8fb54a1c8fd6fe058b18","grade":false,"grade_id":"cell-b5d21fd14b6ed59b","locked":true,"schema_version":3,"solution":false,"task":false},"tags":[],"id":"AvqCB85ID9CX"},"source":["<br>\n","\n","----------------\n","\n","<div style=\"background: rgb(255,100,100); border: solid 1px rgb(129,199,132); padding: 10px;\">\n","<h3>Question 6 (Optional Bonus Question)</h3>\n","</div>\n","\n","> **WARNING: EXTREME DIFFICULTY**<br>\n","> This question is optional, and frankly, not advised.<br>\n","> If attempting this question, be prepared to spend multiple hours or even days.<br>\n","> Successfully completing this question awards bonus marks which can be used to fill-in lost marks in other questions (including the short answer section). <br>\n","> God have mercy on any soul who ventures past this point. <br>\n","\n","<br>\n","\n","In Question 5 we implemented a simplified version of the bubble collapsing process. <br>\n","We deliberately ignored situations such as b) and c) in the figure below, where existing nodes on the consensus path need to be split. <br>\n","Your task is to re-implement the bubble collapsing step to handle these situations.\n","\n","<img src=\"https://raw.githubusercontent.com/melbournebioinformatics/COMP90014_2024/master/assignments/media/collapse_hard.png\" width=\"550\">\n","\n","Hint (from the Velvet publication). <br>\n","\"\"\"<br>\n","*To merge two paths, Tour Bus creates a chain of markers along both of them, node by node. The paths are merged progressively from one end to the next. At each step, the first unmapped minority node is compared to the corresponding majority consensus node, using the local sequence alignment produced previously. All the information attached to that node, including coverage, sequence identifiers, and arcs, is then mapped accordingly onto the majority node. The presence of markers allows Tour Bus to dynamically modify the marked path as it corrects the graph.*<br>\n","\"\"\"\n","\n","\n"]},{"cell_type":"markdown","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"f8f15cc51965450a55eeb68c14dc23a5","grade":false,"grade_id":"cell-dbd5b6121805bc98","locked":true,"schema_version":3,"solution":false,"task":false},"tags":[],"id":"TYdrpCcPD9CX"},"source":["<div style=\"color: rgb(27,94,32); background: rgb(200,230,201); border: solid 1px rgb(129,199,132); padding: 10px;\">\n","<h3>Question 6</h3>\n","\n","<b>Challenge:</b> Write a function which collapses the graph bubble, handling the situations above. <br>\n","\n","- [ ] Input: G (networkx.DiGraph); consensus_path (list of str); minority_path (list of str).\n","- [ ] Ouput: Updated graph with the bubble removed (networkx.DiGraph)\n","    \n","</div>"]},{"cell_type":"code","execution_count":null,"metadata":{"deletable":false,"nbgrader":{"cell_type":"code","checksum":"c05fbef254a70df75ba82b876c6b8f24","grade":false,"grade_id":"cell-c03e72d9ece0c5bf","locked":false,"schema_version":3,"solution":true,"task":false},"tags":[],"id":"0VR4RveOD9CX"},"outputs":[],"source":["# -- OPTIONAL GRADED CELL (3 bonus marks) - complete this cell (if you dare) --\n","\n","def collapse_hard(G: nx.DiGraph, consensus_path: List[str], minority_path: List[str]) -> List:\n","    \"\"\"\n","    Collapses a bubble in flexible manner.\n","    Returns the simplified graph G with bubble removed.\n","    \"\"\"\n","    # YOUR CODE HERE\n","    raise NotImplementedError()\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"jLDIT4TMD9CX"},"outputs":[],"source":["# extra code cell for development if needed\n"]},{"cell_type":"code","execution_count":null,"metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"code","checksum":"a376f945b8da4a34189dc2ca37f3a35c","grade":true,"grade_id":"cell-3c120a22f1e396d7","locked":true,"points":3,"schema_version":3,"solution":false,"task":false},"tags":[],"id":"aXjPn8gCD9CX"},"outputs":[],"source":["# Testing cell - DO NOT ALTER.\n","k = 5\n","G = load_graph(NODELIST, EDGELIST)\n","\n","##############################################################################\n","### Modifying previous graph to emulate situation b) in the figure above.  ###\n","##############################################################################\n","# new 'BC' node\n","l_kmer = G.edges['A', 'B']['kmer']\n","r_kmer = G.edges['C', 'D']['kmer']\n","node_bc_seq = G.nodes['B']['sequence'][:-k+2] + G.nodes['C']['sequence']\n","# remove nodes\n","for node in ['B', 'C', 'C``D`', 'E']:\n","    G.remove_node(node)\n","# add new 'BC' node and edges\n","G.add_node('BC', sequence=node_bc_seq)\n","G.add_edge('A', 'BC', kmer=l_kmer)\n","G.add_edge('BC', 'D', kmer=r_kmer)\n","\n","####################\n","### Single test. ###\n","####################\n","consensus_path = ['A', 'BC', 'D']\n","minority_path = ['A', 'B`', 'C`', 'D']\n","G = collapse_hard(G, consensus_path, minority_path)\n","\n","print(\"expected nodes: ['AB', 'CD', 'X']\")\n","print(f\"actual nodes:   {sorted(G.nodes())}\")\n","edge_ok = True if G.has_edge('AB', 'CD') and G.edges['AB', 'CD']['kmer'] == 'look_' else False\n","print(f\"found edge AB->CD with kmer='look_': {edge_ok}\")\n"]},{"cell_type":"markdown","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"8597a198807560eb3bb667c99b0cb506","grade":false,"grade_id":"cell-d3647d92c359e6e4","locked":true,"schema_version":3,"solution":false,"task":false},"tags":[],"id":"gLa7cJdGD9CX"},"source":["<div style=\"color: rgb(0,0,0); background: #0096FF; border: solid 1px rgb(0,0,0); padding: 10px;\">\n","\n","<h1>Section 2: Short-Answer Questions</h1>\n","\n","</div>\n"]},{"cell_type":"markdown","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"0fc531d59c077888d8c9fa7425c99c7c","grade":false,"grade_id":"cell-ed7f95587d72cb40","locked":true,"schema_version":3,"solution":false,"task":false},"tags":[],"id":"rgLxl2ErD9CX"},"source":["### Introduction\n","\n","Your lab group is investigating protein abundance across different tissues extracted from Barramundi fish. <br>\n","1000 samples were obtained and analysed, covering all major organs, albeit in an unbalanced manner (some organs have more samples than others). <br>\n","You plan to use dimensionality reduction and clustering to visualise and interpret the data, then report any findings to the scientific community.\n","\n"]},{"cell_type":"markdown","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"8d465b050087aaafe5bf1fca2c94323e","grade":false,"grade_id":"cell-c415834e3e86bd75","locked":true,"schema_version":3,"solution":false,"task":false},"tags":[],"id":"TQkrFQ9aD9CX"},"source":["<div style=\"color: rgb(27,94,32); background: rgb(200,230,201); border: solid 1px rgb(129,199,132); padding: 10px;\">\n","<h3>Question 7</h3>\n","\n","(Short Answer Question)<br>\n","(2 marks, max 300 words)\n","\n","<div class=\"alert alert-info\">\n","\n","**Visualising the data**\n","\n","Approach [1 mark]\n","- Outline an approach, incorporating a dimensionality reduction technique (eg PCA, t-SNE, UMAP), to visualize the protein abundance profile of these samples.\n","- Write your answer using dot-point notation (rather than pseudocode) documenting all major steps.\n","\n","Description [1 mark]\n","- Comment on your approach, explaining how common issues associated with your strategy have been handled, and how the choice of parameters may affect the output visualisation.\n","- Write your answer as a single paragraph.\n","\n","</div>\n"]},{"cell_type":"markdown","metadata":{"deletable":false,"nbgrader":{"cell_type":"markdown","checksum":"92637c1852432b9d182845e81ebe8865","grade":true,"grade_id":"cell-5e85fad8f3ca9ca8","locked":false,"points":2,"schema_version":3,"solution":true,"task":false},"tags":[],"id":"0amYfMEBD9CY"},"source":["YOUR ANSWER HERE"]},{"cell_type":"markdown","source":["Approach\n","\n","*   Normalize or scale protein abundance values across samples to ensure uniformity and comparability.\n","*   Apply a dimensionality reduction technique such as PCA to the normalized dataset to condense the high-dimensional data into the top 2-3 components that capture most of the variance.\n","*   Create a 2D or 3D scatter plot using the principal components obtained from PCA, where each point represents a sample. Use different colors or markers to represent different tissue types.\n","*   Optionally, overlay clustering results (e.g., from K-means) on the plot to visualize groups of samples with similar protein abundance profiles.\n","*   Label axes with the percentage of variance explained by each principal component to help interpret the visualized data patterns.\n","\n","Description\n","\n"," The visualization approach employs PCA to create a reduced representation of the high-dimensional dataset, making it easier to identify patterns. By visualizing the top components, we capture the most important variance in the data, simplifying interpretation. The choice of PCA helps reduce noise and highlight key structures, while axis labels provide context for data spread. The output scatter plot offers a visual summary of the dataset, revealing clusters, outliers, and potential correlations between tissue samples based on protein abundance."],"metadata":{"id":"JSvCJjTQzbEp"}},{"cell_type":"markdown","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"e29146ff307084eaae6a38ead36071ee","grade":false,"grade_id":"cell-1db3d59f206147d3","locked":true,"schema_version":3,"solution":false,"task":false},"tags":[],"id":"RXlOUg8AD9CY"},"source":["<div style=\"color: rgb(27,94,32); background: rgb(200,230,201); border: solid 1px rgb(129,199,132); padding: 10px;\">\n","<h3>Question 8</h3>\n","\n","(Short Answer Question)<br>\n","(3 marks, max 300 words)\n","\n","<div class=\"alert alert-info\">\n","\n","**Clustering**\n","    \n","Approach [2 marks]\n","- Outline an approach, using an unsupervised clustering algorithm, to identify tissues which have similar protein abundance profiles.\n","- Write your answer using dot-point notation (rather than pseudocode) documenting all major steps.\n","\n","Description [1 mark]\n","- Indicate whether the raw data or a dimensionality-reduced representation of the data would be used as input to your approach.\n","- Comment on how outliers (eg caused by microbial contamination of samples) or human error (eg mislabeled samples) are handled.\n","- Describe the outputs of your approach, indicating how they can be interpreted to satisfy our goal.\n","- Write your answer as a single paragraph.\n","\n","</div>\n"]},{"cell_type":"markdown","metadata":{"deletable":false,"nbgrader":{"cell_type":"markdown","checksum":"5e1d00de4f896ab0f1975ee8781b512d","grade":true,"grade_id":"cell-060eb7aa17020ef1","locked":false,"points":3,"schema_version":3,"solution":true,"task":false},"tags":[],"id":"lBLPy1prD9CY"},"source":["YOUR ANSWER HERE"]},{"cell_type":"markdown","source":["Approach\n","*   Normalize or scale the protein abundance values to remove the impact of differing scales among features.\n","*   Apply a dimensionality reduction technique like PCA to simplify the dataset and highlight significant patterns, reducing the noise.\n","*   Choose an unsupervised clustering algorithm:\n","\t•\tUse K-means clustering if you have an estimated number of clusters, utilizing the elbow method to determine the optimal cluster count.\n","\t•\tUse DBSCAN to find clusters of varying densities and handle outliers effectively without requiring a preset number of clusters.\n","*   Apply the selected clustering algorithm to the reduced data to identify groups of tissue samples with similar protein abundance profiles.\n","*   Visualize the clusters using scatter plots (for K-means) or density-based plots (for DBSCAN), employing different colors to represent each cluster.\n","*   Analyze cluster composition and potential outliers to draw insights into tissue similarities.\n","\n","Description\n","\n","This clustering approach uses a dimensionality-reduced representation of the data, such as PCA, to streamline clustering and focus on the most informative features. Outliers, caused by contamination or mislabeling, are addressed naturally by DBSCAN, which marks sparse regions as noise, or minimized in K-means by using centroid-based grouping. The clustering output reveals groups of tissue samples with similar protein abundance, which can be interpreted to explore tissue-specific protein patterns, helping to achieve the study’s goals of understanding tissue similarities and variations."],"metadata":{"id":"dYYtnjabH-5W"}},{"cell_type":"markdown","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"5d453868f3bf399a050fc276042940c1","grade":false,"grade_id":"cell-e6403a4b3158d719","locked":true,"schema_version":3,"solution":false,"task":false},"tags":[],"id":"NT5xj0zfD9CY"},"source":["<div style=\"background: rgb(255,165,0); border: solid 1px rgb(129,199,132); padding: 10px;\">    \n","\n","<h1>End of Assignment.</h1>\n","    \n","</div>"]}],"metadata":{"kernelspec":{"display_name":"Python 3 (ipykernel)","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.11.5"},"colab":{"provenance":[]}},"nbformat":4,"nbformat_minor":0}